From 1e8fc71d2d15c2014fbd7f9aa61a20a8e277a4f6 Mon Sep 17 00:00:00 2001
From: Veerendranath Jakkam <quic_vjakkam@quicinc.com>
Date: Thu, 4 Jul 2024 15:11:16 +0530
Subject: [PATCH] nl80211: Add a capability flag for RSN overriding

Add a new capability flag based on the nl80211 feature advertisement
for RSN overriding support.

Patch-dependency: None

Signed-off-by: Veerendranath Jakkam <quic_vjakkam@quicinc.com>
Signed-off-by: Arunpandi Kannan <quic_arunpand@quicinc.com>
---
 src/drivers/driver.h              | 2 ++
 src/drivers/driver_nl80211.c      | 3 +++
 src/drivers/driver_nl80211_capa.c | 6 ++++++
 3 files changed, 11 insertions(+)

diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index 2eaf931..65336a6 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -2434,6 +2434,8 @@ struct wpa_driver_capa {
  */
 #define WPA_DRIVER_FLAGS2_PROT_RANGE_NEG_STA	0x0000000000002000ULL
 #define WPA_DRIVER_FLAGS2_BEACON_RATE_EHT	0x0000000000004000ULL
+/** Driver supports RSN override elements */
+#define WPA_DRIVER_FLAGS2_RSN_OVERRIDE_STA     0x0000000000400000ULL
 /** Driver supports MLO in station/AP mode */
 #define WPA_DRIVER_FLAGS2_MLO			0x0000000000004000ULL
 	u64 flags2;
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index 016f46e..48474d4 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -10105,6 +10105,9 @@ static int nl80211_set_param(void *priv, const char *param)
 			WPA_DRIVER_FLAGS2_SEC_LTF_AP;
 	}
 
+	if (os_strstr(param, "rsn_override_in_driver=1"))
+		drv->capa.flags2 |= WPA_DRIVER_FLAGS2_RSN_OVERRIDE_STA;
+
 	return 0;
 }
 
diff --git a/src/drivers/driver_nl80211_capa.c b/src/drivers/driver_nl80211_capa.c
index a00bf39..fb5da91 100644
--- a/src/drivers/driver_nl80211_capa.c
+++ b/src/drivers/driver_nl80211_capa.c
@@ -1449,6 +1449,12 @@ static void qca_nl80211_get_features(struct wpa_driver_nl80211_data *drv)
 	if (check_feature(QCA_WLAN_VENDOR_FEATURE_AP_ALLOWED_FREQ_LIST,
 			  &info))
 		drv->qca_ap_allowed_freqs = 1;
+	if (check_feature(QCA_WLAN_VENDOR_FEATURE_RSN_OVERRIDE_STA, &info)) {
+		wpa_printf(MSG_DEBUG,
+			   "The driver supports RSN overriding in STA mode");
+			   drv->capa.flags2 |= WPA_DRIVER_FLAGS2_RSN_OVERRIDE_STA;
+	}
+
 	os_free(info.flags);
 }
 
-- 
2.34.1

